capsules/src/signbus/port_layer.rs

```diff
@@ -0,0 +1,382 @@
+               debug_led: Option<&'a hil::gpio::Pin>)
+               -> SignbusPortLayer<'a, A> {
+
+        SignbusPortLayer {
+            i2c: i2c,
+            i2c_send: TakeCell::new(i2c_send),
+            i2c_recv: TakeCell::new(i2c_recv),
+            mod_in_pin: mod_in_pin,
+            mod_out_pin: mod_out_pin,
+            alarm: alarm,
+            debug_led: Cell::new(debug_led),
+            init_client: Cell::new(None),
+            io_client: Cell::new(None),
+            listening: Cell::new(false),
+            master_action: Cell::new(support::MasterAction::Write),
+        }
+    }
+
+    pub fn set_io_client(&self, client: &'static io_layer::SignbusIOLayer) -> ReturnCode {
+        self.io_client.set(Some(client));
+        ReturnCode::SUCCESS
+    }
+
+    pub fn set_init_client(&self,
+                           client: &'static test_signbus_init::SignbusInitialization)
+                           -> ReturnCode {
+        self.init_client.set(Some(client));
+        ReturnCode::SUCCESS
+    }
+}
+
+impl<'a, A: hil::time::Alarm + 'a> PortLayer for SignbusPortLayer<'a, A> {
+    // Set address for this device
+    fn init(&self, i2c_address: u8) -> ReturnCode {
+
+        if i2c_address > 0x7f {
+            return ReturnCode::EINVAL;
+        }
+        hil::i2c::I2CSlave::set_address(self.i2c, i2c_address);
+        ReturnCode::SUCCESS
+    }
+
+    // Do a write to another I2C device
+    fn i2c_master_write(&self, i2c_address: u8, packet: support::Packet, len: usize) -> ReturnCode {
+        self.master_action.set(support::MasterAction::Write);
+
+        self.i2c_send.take().map(|buf| {
+            // packet -> buffer
+            support::serialize_packet(packet, len - support::HEADER_SIZE, buf);
+
+            // Very hacky... add HMAC and change length
+            // Should do this in protocol_layer, but no HMAC in rust
+            buf[5] = 0x2C;
+            buf[12] = 0x96;
+            buf[13] = 0xD3;
+            buf[14] = 0x61;
+            buf[15] = 0x32;
+            buf[16] = 0xEB;
+            buf[17] = 0x6B;
+            buf[18] = 0x0E;
+            buf[19] = 0xBD;
+            buf[20] = 0x09;
+            buf[21] = 0xF0;
+            buf[22] = 0xE4;
+            buf[23] = 0xE5;
+            buf[24] = 0x69;
+            buf[25] = 0x19;
+            buf[26] = 0xA6;
+            buf[27] = 0x3F;
+            buf[28] = 0x78;
+            buf[29] = 0xE2;
+            buf[30] = 0xD6;
+            buf[31] = 0xDE;
+            buf[32] = 0xC8;
+            buf[33] = 0xD5;
+            buf[34] = 0x9A;
+            buf[35] = 0xCA;
+            buf[36] = 0x08;
+            buf[37] = 0x28;
+            buf[38] = 0x9A;
+            buf[39] = 0x17;
+            buf[40] = 0x49;
+            buf[41] = 0xD2;
+            buf[42] = 0xF2;
+            buf[43] = 0xB3;
+
+            hil::i2c::I2CMaster::enable(self.i2c);
+            hil::i2c::I2CMaster::write(self.i2c, i2c_address, buf, 44 as u8);
```

Comment:
Could you comment this all with how it "should be" when the HMAC is done in the protocol layer? Is it just removing the buf writing and changing the size to `buf.len()`?